activesupport/lib/active_support/cache.rb

```diff
@@ -350,6 +341,80 @@ def fetch(name, options = nil, &block)
+      # method. Instead it is responsible for updating the cache out-of-band. It is
+      # important to note that if the block does not update the cache, any existing
+      # value will continously have its expiry pushed forward and thus never expire.
+      #
+      #   cache.write('today', 'Monday', expires_in: 5.minutes)
+      #
+      #   # 6 minutes later:
+      #   cache.fetch_with_deferred_update('today', 1.minutes) {}  # => "Monday"
+      #
+      #   # 2 minutes later:
+      #   cache.fetch_with_deferred_update('today', 1.minutes) {}  # => "Monday"
+      #
+      # The most basic approach would be to use `write` in the block, but that would
+      # normally be a case better handled by `fetch`. Instead, the recommended way
+      # to use the block is to start a background job for expensive calculations and
+      # cache the updated value at the end of that job.
+      #
+      #   cache.fetch_with_deferred_update('expensive_value', 5.minutes) do |key|
+      #     HeavyComputationJob.perform_later key
+      #   end
+      #
+      # While the value is being calculated, the old value will be returned.
+      # The duration should be set so that the calculation has ample time to
+      # complete, to avoid a lot of duplicate cache updates.
+      #
+      # You may specify additional options via the +options+ argument.
+      # Setting <tt>force: true</tt> forces a cache "miss," meaning we treat
+      # the cache value as missing even if it's present. Passing a block is
+      # required when +force+ is true so this always results in a cache write.
+      #
+      #   cache.write('today', 'Monday')
+      #   cache.fetch('today', force: true) { 'Tuesday' } # => 'Tuesday'
+      #   cache.fetch('today', force: true) # => ArgumentError
+      #
+      # You may also specify additional options via the +options+ argument.
+      # Setting <tt>:version</tt> verifies the cache stored under <tt>name</tt>
+      # is of the same version. nil is returned on mismatches despite contents.
+      # This feature is used to support recyclable cache keys.
+      #
+      # Other options will be handled by the specific cache store implementation.
+      # Internally, #fetch_with_deferred_update calls #read_entry, and calls
+      # #write_entry on a cache miss. +options+ will be passed to the #read and
+      # #write calls.
+      def fetch_with_deferred_update(name, additional_cache_time, options = nil, &block)
+        unless block_given?
+          raise ArgumentError, "Missing block: Calling `Cache#fetch_with_deferred_update` requires a block."
+        end
+
+        options = merged_options(options)
+        entry = fetch_entry(name, options, :fetch_with_deferred_update, clear_expired_entry: false)
+
+        if !entry
+          yield(name)
+        elsif entry.expired?
+          write(name, entry.value, options.merge(expires_in: additional_cache_time))
```

Comment:
if you were to pass `race_condition_ttl: additional_cache_time ` to `fetch_entry`, you'd get the extension logic already.

Reply:
@casperisfine Won't it be a slightly different behavior? I wrote the following comment a few days ago:

> I'm not necessarily against using race_condition_ttl, but my initial thought was a slightly different behavior than that option. For this case, I'm thinking each time the block is called, the cached value simply gets a new expiry time. If I read the code right (which is a semi-big if TBH), that is how race_condition_ttl will work for most cache adapters, except for the Redis cache, where there seems to be a "hard expire" once the TTL is reached.

Is my interpretation correct? If so, I'm wondering if it's better for this method to always do the same thing, no matter which adapter is used, rather than to opaquely use an option which is handled slightly differently between implementations.

What do you think?

Reply:
> except for the Redis cache, where there seems to be a "hard expire" once the TTL is reached.

Hum, I can't see such difference. What makes you think that?

Reply:
@casperisfine My bad! 🙈 I had another look at the code and honestly can't remember what made me so sure...

Reply:
@casperisfine Done :+1: